File: daytona-terminal/backend/src/api/websocket.ts
import type { FastifyPluginAsync } from 'fastify';
import type { WebSocket } from 'ws';
import { terminalService } from '../services/terminal.service.js';
import { aiService } from '../services/ai.service.js';
import { createLogger } from '../utils/logger.js';
import { config_ } from '../utils/config.js';

const logger = createLogger('websocket');

interface WSMessage {
  type: 'input' | 'resize' | 'ping' | 'analyze' | 'complete';
  data?: string;
  cols?: number;
  rows?: number;
}

interface WSOutMessage {
  type: 'output' | 'connected' | 'disconnected' | 'error' | 'pong' | 'block' | 'suggestion' | 'completions';
  data?: string;
  sessionId?: string;
  message?: string;
  block?: unknown;
  suggestion?: { text: string; commands: string[] };
  completions?: string[];
}

export const websocketRoutes: FastifyPluginAsync = async (fastify) => {
  // Terminal WebSocket
  fastify.get('/terminal/:sessionId', { websocket: true }, async (socket: WebSocket, request) => {
    const { sessionId } = request.params as { sessionId: string };
    
    logger.info({ sessionId }, 'WebSocket connection attempt');

    // Verify session exists
    const session = await terminalService.getSession(sessionId);
    if (!session) {
      const errorMsg: WSOutMessage = { type: 'error', message: 'Session not found' };
      socket.send(JSON.stringify(errorMsg));
      socket.close(4004, 'Session not found');
      return;
    }

    // Send connected message
    const connectedMsg: WSOutMessage = { type: 'connected', sessionId };
    socket.send(JSON.stringify(connectedMsg));
    logger.info({ sessionId }, 'WebSocket connected');

    // Subscribe to terminal output
    const unsubscribeOutput = terminalService.onOutput(sessionId, (data) => {
      if (socket.readyState === socket.OPEN) {
        const msg: WSOutMessage = { type: 'output', data };
        socket.send(JSON.stringify(msg));
      }
    });

    // Subscribe to command blocks (for error detection)
    const unsubscribeBlocks = terminalService.onBlock(sessionId, (block) => {
      if (socket.readyState === socket.OPEN) {
        const msg: WSOutMessage = { type: 'block', block };
        socket.send(JSON.stringify(msg));

        // If block has error, send auto-suggestion
        if (block.analysis?.hasError && config_.features.autoFix) {
          aiService.quickAnalyze(sessionId, block.output).then((result) => {
            if (result && socket.readyState === socket.OPEN) {
              const suggestionMsg: WSOutMessage = {
                type: 'suggestion',
                suggestion: { text: result.suggestion, commands: result.commands },
              };
              socket.send(JSON.stringify(suggestionMsg));
            }
          }).catch(() => {});
        }
      }
    });

    // Heartbeat
    const heartbeatInterval = setInterval(() => {
      if (socket.readyState === socket.OPEN) {
        socket.ping();
      }
    }, config_.websocket.heartbeatInterval);

    // Handle messages
    socket.on('message', async (rawData) => {
      try {
        const message: WSMessage = JSON.parse(rawData.toString());

        switch (message.type) {
          case 'input':
            if (message.data) {
              await terminalService.write(sessionId, message.data);
            }
            break;

          case 'resize':
            if (message.cols && message.rows) {
              await terminalService.resize(sessionId, message.cols, message.rows);
            }
            break;

          case 'ping':
            socket.send(JSON.stringify({ type: 'pong' }));
            break;

          case 'analyze':
            // Manual analyze request
            if (message.data) {
              const result = await aiService.quickAnalyze(sessionId, message.data);
              if (result) {
                const msg: WSOutMessage = {
                  type: 'suggestion',
                  suggestion: { text: result.suggestion, commands: result.commands },
                };
                socket.send(JSON.stringify(msg));
              }
            }
            break;

          case 'complete':
            // Get AI completions
            if (config_.features.predictiveCommands) {
              const completions = await aiService.getCompletions(sessionId, message.data || '');
              const msg: WSOutMessage = { type: 'completions', completions };
              socket.send(JSON.stringify(msg));
            }
            break;

          default:
            logger.warn({ type: message.type }, 'Unknown message type');
        }
      } catch (error) {
        logger.error({ error }, 'Failed to handle WebSocket message');
      }
    });

    // Handle close
    socket.on('close', () => {
      logger.info({ sessionId }, 'WebSocket disconnected');
      clearInterval(heartbeatInterval);
      unsubscribeOutput();
      unsubscribeBlocks();
    });

    // Handle errors
    socket.on('error', (error) => {
      logger.error({ error, sessionId }, 'WebSocket error');
    });
  });

  // Health check WebSocket
  fastify.get('/health', { websocket: true }, (socket: WebSocket) => {
    socket.send(JSON.stringify({ status: 'ok' }));
    socket.close();
  });
};

File: daytona-terminal/backend/app/api/websocket.py
"""WebSocket handlers for real-time terminal communication."""

import asyncio
import json
from typing import Optional

from fastapi import APIRouter, WebSocket, WebSocketDisconnect

from app.core.logging import logger
from app.services.terminal import get_terminal_service

router = APIRouter(tags=["websocket"])


class TerminalWebSocket:
    """
    WebSocket handler for a terminal session.
    
    Protocol:
    - Incoming messages:
        - {"type": "input", "data": "..."} - Terminal input
        - {"type": "resize", "cols": N, "rows": N} - Resize terminal
        - {"type": "ping"} - Heartbeat ping
    
    - Outgoing messages:
        - {"type": "output", "data": "..."} - Terminal output
        - {"type": "error", "message": "..."} - Error message
        - {"type": "pong"} - Heartbeat pong
        - {"type": "connected"} - Connection established
        - {"type": "disconnected"} - Connection closed
    """

    def __init__(self, websocket: WebSocket, session_id: str):
        self.websocket = websocket
        self.session_id = session_id
        self.terminal_service = get_terminal_service()
        self._closed = False
        self._read_task: Optional[asyncio.Task] = None

    async def connect(self):
        """Accept WebSocket connection."""
        await self.websocket.accept()
        logger.info(f"WebSocket connected for session: {self.session_id}")
        
        # Verify session exists
        session = await self.terminal_service.get_session(self.session_id)
        if not session:
            await self.send_error("Session not found")
            await self.websocket.close(code=4004, reason="Session not found")
            return False
        
        await self.send_message({"type": "connected", "session_id": self.session_id})
        return True

    async def send_message(self, message: dict):
        """Send a JSON message."""
        if not self._closed:
            try:
                await self.websocket.send_json(message)
            except Exception as e:
                logger.warning(f"Failed to send message: {e}")

    async def send_output(self, data: bytes):
        """Send terminal output."""
        if data:
            await self.send_message({
                "type": "output",
                "data": data.decode("utf-8", errors="replace"),
            })

    async def send_error(self, message: str):
        """Send an error message."""
        await self.send_message({"type": "error", "message": message})

    async def handle_input(self, data: str):
        """Handle terminal input."""
        try:
            await self.terminal_service.write_to_terminal(
                self.session_id, data.encode("utf-8")
            )
        except Exception as e:
            await self.send_error(f"Input error: {e}")

    async def handle_resize(self, cols: int, rows: int):
        """Handle terminal resize."""
        try:
            await self.terminal_service.resize_terminal(
                self.session_id, cols, rows
            )
        except Exception as e:
            await self.send_error(f"Resize error: {e}")

    async def start_read_loop(self):
        """Start reading from terminal and sending to WebSocket."""
        async def output_callback(data: bytes):
            await self.send_output(data)
        
        self._read_task = await self.terminal_service.start_read_loop(
            self.session_id, output_callback
        )

    async def run(self):
        """Main WebSocket loop."""
        if not await self.connect():
            return
        
        # Start terminal read loop
        await self.start_read_loop()
        
        try:
            while not self._closed:
                try:
                    # Receive message
                    raw_message = await self.websocket.receive()
                    
                    if raw_message["type"] == "websocket.disconnect":
                        break
                    
                    if "text" in raw_message:
                        message = json.loads(raw_message["text"])
                    elif "bytes" in raw_message:
                        # Binary data is treated as raw input
                        await self.handle_input(
                            raw_message["bytes"].decode("utf-8", errors="replace")
                        )
                        continue
                    else:
                        continue
                    
                    msg_type = message.get("type")
                    
                    if msg_type == "input":
                        await self.handle_input(message.get("data", ""))
                    
                    elif msg_type == "resize":
                        cols = message.get("cols", 120)
                        rows = message.get("rows", 40)
                        await self.handle_resize(cols, rows)
                    
                    elif msg_type == "ping":
                        await self.send_message({"type": "pong"})
                    
                    else:
                        logger.warning(f"Unknown message type: {msg_type}")
                
                except WebSocketDisconnect:
                    break
                except json.JSONDecodeError as e:
                    await self.send_error(f"Invalid JSON: {e}")
                except Exception as e:
                    logger.error(f"WebSocket error: {e}")
                    await self.send_error(str(e))
        
        finally:
            await self.close()

    async def close(self):
        """Clean up connection."""
        if self._closed:
            return
        
        self._closed = True
        
        # Cancel read task
        if self._read_task:
            self._read_task.cancel()
            try:
                await self._read_task
            except asyncio.CancelledError:
                pass
        
        logger.info(f"WebSocket disconnected for session: {self.session_id}")
        
        try:
            await self.send_message({"type": "disconnected"})
        except Exception:
            pass


@router.websocket("/ws/terminal/{session_id}")
async def terminal_websocket(websocket: WebSocket, session_id: str):
    """WebSocket endpoint for terminal I/O."""
    handler = TerminalWebSocket(websocket, session_id)
    await handler.run()


@router.websocket("/ws/health")
async def health_websocket(websocket: WebSocket):
    """Health check WebSocket endpoint."""
    await websocket.accept()
    await websocket.send_json({"status": "ok"})
    await websocket.close()
